metricflow/test/dataflow/optimizer/source_scan/test_source_scan_optimizer.py

```diff
@@ -0,0 +1,337 @@
+from metricflow.test.plan_utils import assert_plan_snapshot_text_equal
+from metricflow.time.time_granularity import TimeGranularity
+
+logger = logging.getLogger(__name__)
+
+
+class ReadSqlSourceNodeCounter(Generic[SourceDataSetT], DataflowPlanNodeVisitor[SourceDataSetT, int]):
+    """Counts the number of ReadSqlSourceNodes in the dataflow plan."""
+
+    def _sum_parents(self, node: DataflowPlanNode[SourceDataSetT]) -> int:
+        return sum(parent_node.accept(self) for parent_node in node.parent_nodes)
+
+    def visit_source_node(self, node: ReadSqlSourceNode[SourceDataSetT]) -> int:  # noqa: D
+        return 1
+
+    def visit_join_to_base_output_node(self, node: JoinToBaseOutputNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_join_aggregated_measures_by_groupby_columns_node(  # noqa: D
+        self, node: JoinAggregatedMeasuresByGroupByColumnsNode[SourceDataSetT]
+    ) -> int:
+        return self._sum_parents(node)
+
+    def visit_aggregate_measures_node(self, node: AggregateMeasuresNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_compute_metrics_node(self, node: ComputeMetricsNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_order_by_limit_node(self, node: OrderByLimitNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_where_constraint_node(self, node: WhereConstraintNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_write_to_result_dataframe_node(self, node: WriteToResultDataframeNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_write_to_result_table_node(self, node: WriteToResultTableNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_pass_elements_filter_node(self, node: FilterElementsNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_combine_metrics_node(self, node: CombineMetricsNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_constrain_time_range_node(self, node: ConstrainTimeRangeNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_join_over_time_range_node(self, node: JoinOverTimeRangeNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_semi_additive_join_node(self, node: SemiAdditiveJoinNode[SourceDataSetT]) -> int:  # noqa: D
+        return self._sum_parents(node)
+
+    def visit_metric_time_dimension_transform_node(  # noqa: D
+        self, node: MetricTimeDimensionTransformNode[SourceDataSetT]
+    ) -> int:
+        return self._sum_parents(node)
+
+    def count_source_nodes(self, dataflow_plan: DataflowPlan[SourceDataSetT]) -> int:  # noqa: D
+        return dataflow_plan.sink_output_node.accept(self)
```

Comment:
It really feels like this entire class should just be `len(dataflow_plan.node_list)` or similar, but I guess that interface is really just "you have a sink node" at the moment.